### ISSUE TITLE
[HELD ON #19298] \[$1000] Unchecking done assignment is not reflected on check-mark 

### ISSUE THREAD

**Issue by mvtglobally** [Expensify/App#19153]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Login to NewDot
2. Create a couple of tasks
3. Mark as complete
4. Sign out
5. Sign in
6. Go to parent chat with tasks
7. Click checkbox to Uncheck done assignment

## Expected Result:
Checkbox should be unmarked

## Actual Result:
Checkbox stays marked like assignment is still done (however in fact it is reverted to pending)

## Workaround:
unknown 

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android / native
- [ ] Android / Chrome
- [ ] iOS / native
- [ ] iOS / Safari
- [X] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

**Version Number:** v1.3.15-5
**Reproducible in staging?:** Y 
**Reproducible in production?:** N 
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation 
Found when executing #18806 

https://github.com/Expensify/App/assets/43995119/5b887f07-673b-473f-a904-9fa7074cd178

**Expensify/Expensify Issue URL:**
**Issue reported by:** Applause - Internal team
**Slack conversation:**

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01fc4b3aaa9fc84298</li>
        <li>Upwork Job ID: 1659567080360730624</li>
        <li>Last Price Increase: 2023-05-19</li>
    </ul>
</details>

**Comment by kevinksullivan** [comment]
OOO today for several days, so I am going to reassign.

**Comment by zanyrenney** [comment]
Hey @kevinksullivan new process for managing issues when you go OOO is to keep yourself assigned. Post in growth [here](https://expensify.slack.com/archives/CK49UADB9/p1684182291511799), as well as the SO [here](https://stackoverflowteams.com/c/expensify/questions/8836) and the thread in #mentors [here](https://expensify.slack.com/archives/C010MRJ5GB0/p1683759930633629).

Basically the person you assign to can oversee/work on the issue but when you return if there is still actions to take, you should help :)

**Comment by zanyrenney** [comment]
Yep, reproducablereproducible: 
<img width="909" alt="2023-05-19_15-29-03" src="https://github.com/Expensify/App/assets/56830058/29e826b5-d5b1-4988-ac62-40a9c0f25f25">

**Comment by zanyrenney** [comment]
@marcochavezf can you confirm if this is a backend bug and i'll change the label please to `internal` ? Thanks!

**Comment by zanyrenney** [comment]
bump @marcochavezf curious for your thoughts before amending the labels here.

**Comment by jjcoffee** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Task checkboxes should work normally after signing out.

### What is the root cause of that problem?
When we log back in, the linked `taskReport` doesn't load (doesn't appear in LHN and doesn't show up in the `TaskPreview`'s `props.taskReport` and isn't in Onyx). It looks like this will be fixed in #19090. 

The `reportAction` itself **does** have the correct state of the task, however, so I'd expect the checkbox to still work. `TaskPreview` is getting the task's completion state here:

https://github.com/Expensify/App/blob/c917c447f225fe3b194a5e03f568cca0de9e5e71/src/components/ReportActionItem/TaskPreview.js#L56-L58

I think the intention here is to get the task's state from *either* the `taskReport` itself, or from the `reportAction`, however what we're actually doing is saying the task is complete if either source reports that the task is complete.

The problem with this is that these sources aren't always in sync (in this case it's more to do with the BE issue, but the same could occur if going offline).

In the case of our issue, when you log back in and try and reopen the task the linked `taskReport` correctly gets updated to a reopened state (so the first part of the `isTaskCompleted` expression is false), but `reportAction` doesn't update its `childStateNum` or `childStatusNum` at all, so we have:

- `taskReport` - **OPEN**
- `reportAction` - **COMPLETED**

Which means `isTaskCompleted` always returns true, and we can never "untick" the checkbox.

### What changes do you think we should make in order to solve the problem?
Firstly I think we should treat the `taskReport` as a ground source of truth (where it is available), as it's what the API actually updates in its responses. So we'd want to update this line:

https://github.com/Expensify/App/blob/c917c447f225fe3b194a5e03f568cca0de9e5e71/src/components/ReportActionItem/TaskPreview.js#L56-L58

To something like this:

```
const isTaskCompleted = !_.isEmpty(props.taskReport) ?
	(props.taskReport.stateNum === CONST.REPORT.STATE_NUM.SUBMITTED && props.taskReport.statusNum === CONST.REPORT.STATUS.APPROVED) :
	(props.action.childStateNum === CONST.REPORT.STATE_NUM.SUBMITTED && props.action.childStatusNum === CONST.REPORT.STATUS.APPROVED);
```

Then to deal with the issue of the `reportAction` never updating with the task's completion state, we may want ensure that the BE response includes an update for the `reportAction` when we call either the `ReopenTask` or `CompleteTask` endpoints. We should also add it as both `optimisticData` and `successData` to handle offline cases, e.g. updating this:

https://github.com/Expensify/App/blob/5a3458e9b33bfe3d8aea4a78cb5e2f53568b9186/src/libs/actions/Task.js#L155

To pass in the `reportActionID`:

`function completeTask(taskReportID, parentReportID, reportActionID, taskTitle)`

And then add that to both `optimisticData` and `successData`:

```
{
	onyxMethod: Onyx.METHOD.MERGE,
	key: `${ONYXKEYS.COLLECTION.REPORT_ACTIONS}${parentReportID}`,
	value: {[reportActionID]: {
		childStateNum: CONST.REPORT.STATE_NUM.SUBMITTED,
		childStatusNum: CONST.REPORT.STATUS.APPROVED,
	}},
}
```

We also need to pass the `reportActionID` from the `TaskPreview` itself here:

https://github.com/Expensify/App/blob/0ec61bf1ca783ee0e30fbbc3db7285a33e96744a/src/components/ReportActionItem/TaskPreview.js#L77

`TaskUtils.completeTask(props.taskReportID, parentReportID, props.action.reportActionID, taskTitle);`

The same needs to be done for `reopenTask` (with the relevant report state CONSTs updated).

**Comment by marcochavezf** [comment]
Oh interesting, thanks @jjcoffee for pointing that out! I'm thinking of putting this on hold for https://github.com/Expensify/App/issues/19090, and once it's done, we can come back to implement the FE part here. Alternatively, do you think is better to do the FE changes before the BE part (so we can still reproduce the error)? cc @0xmiroslav

**Comment by kevinksullivan** [comment]
thanks for the heads up @zanyrenney . Missed that update

**Comment by kevinksullivan** [comment]
Waiting on @marcochavezf here

**Comment by marcochavezf** [comment]
Sorry for the delay here. Perfect, let's implement the FE changes, assigning @jjcoffee

**Comment by jjcoffee** [comment]
Thanks @marcochavezf! [PR](https://github.com/Expensify/App/pull/19671) is ready for review @0xmiroslav. Note that there have been some BE changes so now the task title doesn't show after signing out, and also the "Reopened task" message won't get added - this behaviour is unrelated to the changes in this PR.

**Comment by kevinksullivan** [comment]
@jjcoffee @0xmiroslav offers sent in upwork. Thanks!

**Comment by kevinksullivan** [comment]
Tagged @marcochavezf in slack to bump as well

https://expensify.slack.com/archives/C049HHMV9SM/p1685751288985569?thread_ts=1684871320.623729&cid=C049HHMV9SM

**Comment by marcochavezf** [comment]
Commented [here](https://github.com/Expensify/App/pull/19671#issuecomment-1577362484).

**Comment by jjcoffee** [comment]
@kevinksullivan Can you put this on hold for the BE issues that are being sorted in #19298? See @marcochavezf's [comment](https://github.com/Expensify/App/pull/19671#issuecomment-1577362484). I'll then re-raise my PR once this is off hold (it will need retesting mainly in offline mode).

**Comment by kevinksullivan** [comment]
Done @jjcoffee

**Comment by kevinksullivan** [comment]
Still on hold

**Comment by kevinksullivan** [comment]
Held

**Comment by jjcoffee** [comment]
@0xmiroslav I've retested and it is no longer reproducible in the same way. The underlying issue of the `reportAction` `props.action.childStateNum` and `props.action.childStatusNum` not being updated optimistically (or by the BE) is still there, but it does not manifest in practise (from what I could see in my tests). 

It looks like the initial part of my fix (the tweak to `isCompletedTask` was added to [this PR](https://github.com/Expensify/App/pull/20014). This isn't quite correctly implemented, see [this bug](https://expensify.slack.com/archives/C049HHMV9SM/p1687789907864819) I've posted on Slack. We can either handle that here or in a separate issue, I'm not sure what's better?

If we don't handle it here, I'm not sure how this works with compensation as I think technically when hired and assigned, compensation is guaranteed (according to [this](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#:~:text=Compensation%20is%20only%20guaranteed%20to%20those%20who%20propose%20a%20solution%20and%20get%20hired%20for%20that%20job.)). It does take quite a bit of time to put together PRs and do all the testing, so I would expect compensation here. cc @kevinksullivan

**Comment by kevinksullivan** [comment]
Hi @jjcoffee , can you elaborate on what you're proposing? You're saying there is a new, somewhat related bug to the one you reported, and you're looking to fix it? Or something else? Sorry if I am just misreading you here.

**Comment by jjcoffee** [comment]
@kevinksullivan Sure! I discovered a related bug whilst retesting this issue, but the test steps are quite different as it involves two users:

1. Log in to NewDot on Chrome web with User A
2. User A opens a chat with User B and creates a task assigned to User A
3. User A completes the task
4. User B signs in and opens the chat, the checkbox shows the task still open.

The RCA and fix is the same as my proposal here, that's why I thought maybe we can just handle it here.

**Comment by jjcoffee** [comment]
The related issue I mentioned is being handled [here](https://github.com/Expensify/App/issues/21924). Checking [on Slack](https://expensify.slack.com/archives/C01GTK53T8Q/p1688549271658019) regarding compensation in this slightly odd case!

**Comment by kevinksullivan** [comment]
checking internally

https://expensify.slack.com/archives/C01SKUP7QR0/p1689282907772589

**Comment by kevinksullivan** [comment]
Concluded this is not eligible for payout since it wasn't solved in this issue. Going to close out but let me know if you have any further questions on this.
